### ISSUE TITLE
Support for class usage

### ISSUE THREAD

**Issue by tripodsan** [classgraph/classgraph#122]
Would it be possible to also get the dependency information to all the classes that are used by a class (e.g.calling their methods or accessing their fields) ?
this would allow to create a complete dependency tree.

**Comment by lukehutch** [comment]
Good question, you could do it by parsing all the bytecode in a class, but I wonder if the list of dependencies is stored in the classfile by the compiler (I think it is not). There is support for getting all classes with fields of a given type already, which is the inverse of half of your proposal, but doing this for method calls may not be possible without parsing the bytecode. I'm currently traveling, so it may take me some time to research this, but feel free to read through the relevant Java language spec sections to check this, if you have time.

Out of curiosity, what is your use case?

**Comment by tripodsan** [comment]
> Out of curiosity, what is your use case?

- find the dependencies of a set of classes to a given set of libraries during build time. For example embedded in a maven plugin, we could detect the `<dependencies/>` that are not used to compile the artifact.
- find the dependencies between different libraries in oder to improve modularization and create a dependency graph.

**Comment by tripodsan** [comment]
btw, I think the _Analyzer_ of the bnd tools has what I need:
https://github.com/bndtools/bnd/blob/master/biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java

the `Clazz.getReferred()` returns all the imports of a class, which are basically all its dependencies.

**Comment by lukehutch** [comment]
Good find, and yes, that class does seem to do what you need, by passing the bytecode directly. I wasn't able to find any mention of dependencies in the Java spec, so this must be the only way to do it:

http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html

BTW the Analyzer class you linked does not seem to always call crawl() for reasons that I didn't look into closely, so you might want to double-check the logic to ensure you're always crawling the bytecode, if you want to make sure you always get all the dependency refs.
 I'll close this for now, since parsing the bytecode is currently out of scope for FastClasspathScanner, but I will link to bnd from the FCS wiki. Thanks for the report!

**Comment by lukehutch** [comment]
One other thing -- for the case of Maven specifically I believe Maven publishes a modified pom file within a published jar -- that may contain the dependencies you need.

**Comment by t1** [comment]
I would like to reactivate this issue and contribute an optional scanner. What do you think?

**Comment by lukehutch** [comment]
It's mostly possible with the ClassGraph API today, at least for types (and generic type parameters) of classes, annotations, fields, methods and method parameters. Can you please check if there are cases where the only reference to a class is in the bytecode (eg. a static method call or a local variable type)?

Classgraph doesn't touch the bytecode yet. If you want to extend ClassfileBinaryParser to scan bytecode looking for local types, without adding an external dependency on ASM or similar, please go for it!

**Comment by lukehutch** [comment]
@t1 Actually I just realized that every reference to a class name in the bytecode should refer back to the constant pool table (a quick check indicates this is the case for local variables and for static method invocations). So you wouldn't even have to parse the bytecodes. What you would do is:

1. Add a method to `ClassGraph`, e.g. `.findClassDependencies()`, that sets a field in `ScanSpec`, `findClassDependencies`, to `true` (the field should have a default assignment of `false`).
2. Add logic to the constant pool reader to parse class names found in any currently-ignored constant pool entries if `scanSpec.findClassDependencies` is `true`: https://github.com/classgraph/classgraph/blob/master/src/main/java/io/github/classgraph/ClassfileBinaryParser.java#L352
3. Store these class names in a new `Set` as they are encountered (only initialize the `Set` to a new `HashSet` if `scanSpec.findClassDependencies` is `true`).
4. For any non-ignored constant pool entries that contain class names, e.g. supertypes, implemented interfaces, annotations, etc., add those to the `Set` too. However, don't parse generic type signatures, since these are costly and time-consuming to parse (they will be parsed on demand when dependency information is requested).
5. Store the `Set` in a new field in `ClassInfoUnlinked`.
6. When a new `ClassInfo` class is constructed, if `scanSpec.findClassDependencies` is `true`, create a new `HashSet` as a field of `ClassInfo` to store the class dependencies. Then when a `ClassInfoUnlinked` class is linked into (merged into) a `ClassInfo` object, you will need to `addAll` the values from the set in `ClassInfoUnlinked`. (You can see this merging process happening for other `ClassInfoUnlinked` fields -- this is necessary because a `ClassInfo` placeholder object can be created before the class is actually scanned.)
7. In `ClassInfo`, add a method `.getClassDependencies()` that creates a new local `HashSet`, and merges in the class dependencies from the new field in `ClassInfo`, and then:
  1. Call `.getTypeSignatureOrTypeDescriptor()` on the `ClassInfo` itself, and on each superclass, implemented interface or annotation. Check the concrete type of the returned `TypeSignature` (it can be one of several sub-types), cast it to the appropriate subtype, and extract any class references (or generic type parameter references).
  2. For methods, you need to separately get types from the return type of the method and each `MethodParameterInfo` for each method.
  3. For annotations on the class, each field, each method, and each method parameter, you need to check if the annotation contains an annotation parameter with a value that has type `AnnotationClassRef`, and add that referenced class to the set.
8. For all class names in the set, look up the `ClassInfo` for the named class using `scanResult.classNameToClassInfo.get(className)`, and if non-null, add it to a new `Set<ClassInfo>`.
9. Construct a new `ClassInfoList` from the `Set<ClassInfo>` and return it.
10. Add a method to `ScanResult`, `.getAllClassDependencies()`, that gets the dependencies from each class, and returns a `Map<ClassInfo, ClassInfoList>` mapping each `ClassInfo` to its dependencies. Also add method `.getAllDependentClasses()` that is also a `Map<ClassInfo, ClassInfoList>`, but implements a reverse index mapping dependencies to dependent classes.

I think that should be everything. Fortunately there's nothing as difficult as writing a new bytecode parser in this list -- everything should be relatively straightforward. Let me know if you need help with anything specific.

**Comment by t1** [comment]
@lukehutch: Many thanks for the hints... more that I could have hoped for! I'm on it.

**Comment by t1** [comment]
I'm doing some progress (in my free time) and have some questions I'd like to discuss with you:
* If we define 'class dependency' to mean that a class can't run without that other class, then annotations are not included, as a class can perfectly run without the annotation classes it uses on the classpath. It does need annotations, OTOH, when it uses them, e.g., for reflection. Excluding applied annotations is exactly what I need for my use-case. Can you think of any other realistic use-case that would also require dependencies on annotations applied (not accessed)?
* Class references from a method body are not stored with a class tag (7), but only with a string tag (1), so to be able to distinguish them from a string starting with a capital L and ending with a semicolon, I fear we need to completely parse the body, too.

**Comment by lukehutch** [comment]
On your first point: I agree that annotations have somewhat of a special function, and maybe should not be thought of as class dependencies (for example, ClassGraph is compiled in JDK 7 compatibility mode, but needs to be compiled using JDK 8+, because it uses `@FunctionalInterface` annotations on several interfaces, so that lambdas can be used -- JDK 7 doesn't know anything about this annotation class, but simply ignores it if you don't try querying the annotations on those interface classes; in particular, you don't get a `LinkageError` or similar, as you would with a normal class reference). However, for completeness, I think it's probably a much better idea to return all class dependencies, including those in annotations. Users are less likely to run into surprises that way.

On your second point: I didn't realize that class names were sometimes stored in string tags. That would be a problem, if you are hoping to extract class refs from the constant pool. You could get some of the class references by looking at the `LocalVariableTypeTable` and/or the `LocalVariableTable` attribute in the `attributes` table of the `Code` attribute of a classfile:

https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.7.14
https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.7.13

This may miss a few things, such as class constant references (`MyClass.class`), possibly -- I don't know enough about Java bytecodes to know all the ways classes can be referenced by bytecodes, unfortunately...

Maybe it's better to think about usecases for this, and drive the design from there. @t1 / @tripodsan, what specific things do you need this for?

**Comment by hepin1989** [comment]
@t1 Take it easy, balance work ,life and opensourcing.

**Comment by lukehutch** [comment]
> This seems to grow beyond what I can allocate to a purely free time project 😢

That's what I said when I started working on ClassGraph! :D   Now 3 years later... 

No problem, if more people need this, it will get done eventually by someone! I need to be encouraging more contributions like this, since in the past if someone had a feature request, I would just go implement it... 

I'm still curious what the needs and usecases are that drove this feature request, and your interest in working on it. There may be a reduced-size problem that could be solved more simply.

**Comment by t1** [comment]
I'm working on an annotation processor that checks your package dependencies [pdap](https://github.com/t1/pdap). I just found out that I can access the Java AST directly without scanning the class files... which isn't even possible, as they are not yet written when the annotation processor runs!
BTW: I looked at the code of JDepend, and scanning the methods for dependencies isn't actually that bad. See `ClassFileParser.parseMethods`.

**Comment by lukehutch** [comment]
I looked at `ClassFileParser.parseMethods` in JDepend, and all it is doing is looking at the type signature of each method, and extracting type names from the type signature string. It doesn't look at local variables or the types of intermediate expressions.

You could get the same exact information as JDepend from `MethodInfo` and `MethodParameterInfo` in ClassGraph today, without writing any new code. However, to get local variable types, you would need to parse the `LocalVariableTypeTable` (or if not present, the non-generic version, `LocalVariableTable`) in the `attributes` table of the `Code` attribute of a classfile, as mentioned above. It actually wouldn't be very much additional code that you would need to add to `ClassGraph`'s `ClassFileBinaryParser` to get the types of local variables of each method.

However, for the types of intermediate values, you might need to do some more work, probably parsing bytecodes -- but it's not even clear if you'd even want to count the types of intermediate values as a dependency. For example, if you have:

```
class A {
}

class B {
    A getA() { return new A(); }
}

class C {
    B getB() { return new B(); }
}

class D {
    A getA() {
        return new C().getB().getA();
    }
}
```

Do you actually want class `D` to show as having class `C` and class `B` as a dependency? Or only class `A`?

**Comment by t1** [comment]
You're right, `ClassFileParser.parseMethods` doesn't look at local variables! Strangely enough, JDepend *does* find local dependencies! E.g. `ClassFileParserTest.testConcreteClass` finds a dependency on `java.net` which is only used for a method variable. But I don't understand, how they do it.

And yes: We *would* want to have `B` and `C` as dependencies. `D` wouldn't load without them, would it?

**Comment by lukehutch** [comment]
Have you looked at how the OpenJDK classfile verifier finds B and C? That would be interesting.

**Comment by t1** [comment]
> Have you looked at how the OpenJDK classfile verifier finds B and C? That would be interesting.

I would if I could spare some time 😢

**Comment by lukehutch** [comment]
> What caused you to come to this conclusion?:

Ah, there is an extra layer of indirection here. Method and field ref constant pool entries redirect to string constant pool entries. However, they should never be tagged as only a string.

I'll take a shot at implementing the extraction of class refs from the constant pool.

**Comment by lukehutch** [comment]
I implemented this feature, more or less following the steps in my long comment above.

I didn't have to parse the bytecodes -- everything I could find or could think of that could reference a type seemed to be stored in the constant pool table as a class reference or "name and type" reference, or as a type signature string constant. I extracted class names from all of these, as well as for annotations, as well as enum classes and class refs in (possibly nested) annotation parameter values.

@tripodsan @t1 Please test the git master version to see if it catches all corner cases that you think it should catch. (It works for the `A`, `B`, `C`, `D` example I gave above.) Feel free to submit unit tests as a pull request, if you would like.

The methods you can call are:

* `ClassInfoList ClassInfo#getClassDependencies()`
* `Map<ClassInfo, ClassInfoList> ScanResult#getClassDependencyMap()`
* `Map<ClassInfo, ClassInfoList> ScanResult#getReverseClassDependencyMap()`

Please also comment on whether you prefer `getClassDependencies()` or `getReferencedClasses()`.

**Comment by lukehutch** [comment]
Released in 4.6.11. Please comment back here if this doesn't do what you expect. Thanks!
